relays/ethereum/src/ethereum_exchange.rs

```diff
@@ -0,0 +1,211 @@
+use crate::ethereum_types::{
+	EthereumHeaderId, Transaction as EthereumTransaction, TransactionHash as EthereumTransactionHash, H256,
+};
+use crate::exchange::{relay_single_transaction_proof, SourceClient, TargetClient, TransactionProofPipeline};
+use crate::rpc::{EthereumRpc, SubstrateRpc};
+use crate::rpc_errors::{EthereumNodeError, RpcError};
+use crate::substrate_client::{
+	SubmitEthereumExchangeTransactionProof, SubstrateConnectionParams, SubstrateRpcClient, SubstrateSigningParams,
+};
+use crate::sync_types::HeaderId;
+
+use async_trait::async_trait;
+use bridge_node_runtime::exchange::EthereumTransactionInclusionProof;
+use std::time::Duration;
+
+/// Interval at which we ask Ethereum node for updates.
+const ETHEREUM_TICK_INTERVAL: Duration = Duration::from_secs(10);
+/// Interval at which we ask Substrate node for updates.
+const SUBSTRATE_TICK_INTERVAL: Duration = Duration::from_secs(5);
+
+/// PoA exchange transaction relay params.
+#[derive(Debug, Default)]
+pub struct EthereumExchangeParams {
+	/// Ethereum connection params.
+	pub eth: EthereumConnectionParams,
+	/// Hash of the Ethereum transaction to relay.
+	pub eth_tx_hash: EthereumTransactionHash,
+	/// Substrate connection params.
+	pub sub: SubstrateConnectionParams,
+	/// Substrate connection params.
+	pub sub_sign: SubstrateSigningParams,
+}
+
+/// Ethereum to Substrate exchange pipeline.
+struct EthereumToSubstrateExchange;
+
+impl TransactionProofPipeline for EthereumToSubstrateExchange {
+	const SOURCE_NAME: &'static str = "Ethereum";
+	const TARGET_NAME: &'static str = "Substrate";
+
+	type BlockHash = H256;
+	type BlockNumber = u64;
+	type TransactionHash = EthereumTransactionHash;
+	type Transaction = EthereumTransaction;
+	type TransactionProof = EthereumTransactionInclusionProof;
+}
+
+/// Ethereum node as transactions proof source.
+struct EthereumTransactionsSource {
+	client: EthereumRpcClient,
+}
+
+#[async_trait]
+impl SourceClient<EthereumToSubstrateExchange> for EthereumTransactionsSource {
+	type Error = RpcError;
+
+	async fn tick(&self) {
+		async_std::task::sleep(ETHEREUM_TICK_INTERVAL).await;
+	}
+
+	async fn transaction(
+		&self,
+		hash: &EthereumTransactionHash,
+	) -> Result<Option<(EthereumHeaderId, EthereumTransaction)>, Self::Error> {
+		let eth_tx = match self.client.transaction_by_hash(*hash).await? {
+			Some(eth_tx) => eth_tx,
+			None => return Ok(None),
+		};
+
+		// we need transaction to be mined => check if it is included in the block
+		let eth_header_id = match (eth_tx.block_number, eth_tx.block_hash) {
+			(Some(block_number), Some(block_hash)) => HeaderId(block_number.as_u64(), block_hash),
+			_ => return Ok(None),
+		};
+
+		Ok(Some((eth_header_id, eth_tx)))
+	}
+
+	async fn transaction_proof(
+		&self,
+		eth_header_id: &EthereumHeaderId,
+		eth_tx: EthereumTransaction,
+	) -> Result<EthereumTransactionInclusionProof, Self::Error> {
+		let eth_header = self.client.header_by_hash(eth_header_id.1).await?;
```

Comment:
IMHO we should get both with transactions in one RPC query. It would also simplify the code a tiny bit.

Reply:
Done, but the bad thing is that now we have this:
```
		#[rpc(method = "eth_getBlockByHash", positional_params)]
		fn get_block_by_hash(hash: H256, full_tx_objs: bool) -> EthereumHeader;
		#[rpc(method = "eth_getBlockByHash", positional_params)]
		fn get_block_by_hash_with_transactions(hash: H256, full_tx_objs: bool) -> EthereumHeaderWithTransactions;
```

in rpc.rs. So basically if wrong flag is passed, we'll fail to decode results. I do not know any other how to fix that, but to introduce custom deserialization (which I wanted to avoid).

Reply:
Yeah, I think that's fine. The ethereum API is not really well-typed. AFAIR in ParityEthereum we were returning an `enum` and depending on the flag different variant was returned. I'm also fine to return `(EthereumHeader, Option<Vec<Transactions>>>)` to avoid introducing another type, but your solution looks good too.